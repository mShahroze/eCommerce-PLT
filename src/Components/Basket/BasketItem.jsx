import React, { useState } from "react";
import "../Styles/BasketItem.css";
import { connect } from "react-redux";
import {
  removeFromBasket,
  adjustItemQuantity,
} from "../../Redux/Shopping/shopping-actions";

const BasketItem = ({ item, removeFromBasket, adjustQuantity }) => {
  const [input, setInput] = useState(item.quantity);
  const onChangeHandler = (e) => {
    setInput(e.target.value);
    adjustQuantity(item.id, e.target.value);
  };

  return (
    <div data-testid="basketItem" className="basketItem">
      <img className="basketItemImg" src={item.img} alt={item.name} />
      <div className="basketItemDetails">
        <p className="detailsName">{item.name}</p>
        <p className="detailsColour">{item.colour}</p>
        <p className="detailsPrice">Â£ {item.price}</p>
      </div>
      <div className="basketItemActions">
        <div className="basketItemQuantity">
          <label htmlFor="qty">Qty</label>
          <input
            data-testid="input"
            min="1"
            type="number"
            id="qty"
            name="qty"
            value={input}
            onChange={onChangeHandler}
          />
        </div>
        <button
          data-testid="removeItem"
          onClick={() => removeFromBasket(item.id)}
          className="actionsDeleteItemButton"
        >
          <img
            src="https://image.flaticon.com/icons/svg/709/709519.svg"
            alt=""
          />
        </button>
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    adjustQuantity: (id, quantity) =>
      dispatch(adjustItemQuantity(id, quantity)),
    removeFromBasket: (id) => dispatch(removeFromBasket(id)),
  };
};

export default connect(null, mapDispatchToProps)(BasketItem);
