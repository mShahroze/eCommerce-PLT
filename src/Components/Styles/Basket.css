.basket {
  margin: 2rem auto;
  width: 1100px;
  display: flex;
  justify-content: space-between;
}

.basketItems {
  flex: 0.7;
  margin-right: 1rem;
}

.basketSummary {
  height: 250px;
  flex: 0.3;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 1rem;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.26);
  border-radius: 10px;
}

.summaryName {
  font-size: 1.2rem;
  font-weight: bold;
}

.summaryPrice {
  display: flex;
  align-items: center;
}

.summaryPrice span:nth-child(1) {
  font-size: 0.9rem;
  flex: 0.6;
}

.summaryPrice span:nth-child(2) {
  flex: 0.4;
  font-size: 1.2rem;
  font-weight: bold;
}

.summaryCheckoutButton {
  padding: 10px 17px;
  border: 1px solid var(--secondary-color);
  border-radius: 10px;
  background: var(--primary-color);
  color: var(--light-color);
  font-size: 1.2rem;
  font-weight: bold;
  cursor: pointer;
}

.summaryCheckoutButton:focus {
  outline: none;
  border: 1px solid var(--secondary-color);
}

.summaryCheckoutButton:hover {
  opacity: 0.75;
}
