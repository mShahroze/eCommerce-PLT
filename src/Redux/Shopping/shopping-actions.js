import * as actionTypes from "./shopping-types";
import axios from "axios";

export const getProducts = () => async (dispatch) => {
  try {
    const res = await axios.get(
      "https://my-json-server.typicode.com/benirvingplt/products/products"
    );
    dispatch({
      type: actionTypes.GET_PRODUCTS,
      payload: res.data,
    });
  } catch (e) {
    dispatch({
      type: actionTypes.PRODUCTS_ERROR,
      payload: console.log(e),
    });
  }
};

export const addToBasket = (itemID) => {
  return {
    type: actionTypes.ADD_TO_BASKET,
    payload: {
      id: itemID,
    },
  };
};

export const removeFromBasket = (itemID) => {
  return {
    type: actionTypes.REMOVE_FROM_BASKET,
    payload: {
      id: itemID,
    },
  };
};

export const adjustItemQuantity = (itemID, quantity) => {
  return {
    type: actionTypes.ADJUST_ITEM_QUANTITY,
    payload: {
      id: itemID,
      quantity: quantity,
    },
  };
};

export const loadCurrentItem = (item) => {
  return {
    type: actionTypes.LOAD_CURRENT_ITEM,
    payload: item,
  };
};
