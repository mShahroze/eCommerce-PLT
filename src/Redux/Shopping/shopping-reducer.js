import * as actionTypes from "./shopping-types";

const INITIAL_STATE = {
  products: [],
  basket: [],
  currentItem: null,
  loading: true,
};

const shopReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case actionTypes.GET_PRODUCTS:
      return {
        ...state,
        products: action.payload,
        loading: false,
      };
    case actionTypes.ADD_TO_BASKET:
      const item = state.products.find((prod) => prod.id === action.payload.id);
      const inBasket = state.basket.find((item) =>
        item.id === action.payload.id ? true : false
      );
      return {
        ...state,
        basket: inBasket
          ? state.basket.map((item) =>
              item.id === action.payload.id
                ? { ...item, quantity: item.quantity + 1 }
                : item
            )
          : [...state.basket, { ...item, quantity: 1 }],
      };
    case actionTypes.REMOVE_FROM_BASKET:
      return {
        ...state,
        basket: state.basket.filter((item) => item.id !== action.payload.id),
      };
    case actionTypes.ADJUST_ITEM_QUANTITY:
      return {
        ...state,
        basket: state.basket.map((item) =>
          item.id === action.payload.id
            ? { ...item, quantity: +action.payload.quantity }
            : item
        ),
      };
    case actionTypes.LOAD_CURRENT_ITEM:
      return {
        ...state,
        currentItem: action.payload,
      };
    default:
      return state;
  }
};

export default shopReducer;
