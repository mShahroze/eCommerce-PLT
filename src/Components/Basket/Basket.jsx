import React, { useState, useEffect } from "react";
import "../Styles/Basket.css";
import { connect } from "react-redux";
import BasketItem from "./BasketItem";

const Basket = ({ basket }) => {
  const [totalPrice, setTotalPrice] = useState(0);
  const [totalItems, setTotalItems] = useState(0);

  useEffect(() => {
    let items = 0;
    let price = 0;

    basket?.forEach((item) => {
      items += item.quantity;
      price += item.quantity * item.price;
    });

    setTotalItems(items);
    setTotalPrice(price);
  }, [basket, totalPrice, totalItems, setTotalPrice, setTotalItems]);

  return (
    <div className="basket">
      <div data-testid="basket" className="basketItems">
        {basket?.map((item) => (
          <BasketItem key={item.id} item={item} />
        ))}
      </div>
      <div className="basketSummary">
        <h4 className="summaryName">Cart Summary</h4>
        <div className="summaryPrice">
          <span>TOTAL: {totalItems}</span>
          <span>Â£ {totalPrice.toFixed(2)}</span>
        </div>
        <button className="summaryCheckoutButton">Proceed To Checkout</button>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    basket: state.shop.basket,
  };
};

export default connect(mapStateToProps)(Basket);
