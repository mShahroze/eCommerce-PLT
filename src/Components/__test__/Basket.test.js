import React from "react";
import Basket from "../Basket/Basket";
import { fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { applyMiddleware } from "redux";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import rootReducer from "../../Redux/rootReducer";
import { removeFromBasket } from "../../Redux/Shopping/shopping-actions";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  shop: {
    products: [],
    basket: [
      {
        id: 1,
        colour: "Black",
        name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
        price: 10,
        img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
        quantity: 1,
      },
    ],
    currentItem: null,
    loading: true,
  },
};

const store = mockStore(initialState);

describe("Basket component", () => {
  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Basket />
        </BrowserRouter>
      </Provider>
    );

    expect(container);
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Basket />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("Remove item from basket", async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Basket />
        </BrowserRouter>
      </Provider>
    );
    const basketEl = getByTestId("basket");
    expect(basketEl.childElementCount).toBe(1);
    const removeItemEl = getByTestId("removeItem");
    fireEvent.click(removeItemEl);
    const calledActions = store.getActions();
    expect(calledActions[0].type).toBe("REMOVE_FROM_BASKET");
  });
});
