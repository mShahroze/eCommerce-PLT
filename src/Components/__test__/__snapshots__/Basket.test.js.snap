// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basket component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="basket"
  >
    <div
      class="basketItems"
      data-testid="basket"
    >
      <div
        class="basketItem"
        data-testid="basketItem"
      >
        <img
          alt="Black Sheet Strappy Textured Glitter Bodycon Dress"
          class="basketItemImg"
          src="http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024"
        />
        <div
          class="basketItemDetails"
        >
          <p
            class="detailsName"
          >
            Black Sheet Strappy Textured Glitter Bodycon Dress
          </p>
          <p
            class="detailsColour"
          >
            Black
          </p>
          <p
            class="detailsPrice"
          >
            £ 10
          </p>
        </div>
        <div
          class="basketItemActions"
        >
          <div
            class="basketItemQuantity"
          >
            <label
              for="qty"
            >
              Qty
            </label>
            <input
              data-testid="input"
              id="qty"
              min="1"
              name="qty"
              type="number"
              value="1"
            />
          </div>
          <button
            class="actionsDeleteItemButton"
            data-testid="removeItem"
          >
            <img
              alt=""
              src="https://image.flaticon.com/icons/svg/709/709519.svg"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="basketSummary"
    >
      <h4
        class="summaryName"
      >
        Cart Summary
      </h4>
      <div
        class="summaryPrice"
      >
        <span>
          TOTAL: 1
        </span>
        <span>
          £ 10.00
        </span>
      </div>
      <button
        class="summaryCheckoutButton"
      >
        Proceed To Checkout
      </button>
    </div>
  </div>
</DocumentFragment>
`;
